cmd/kubeadm/app/util/users/users_linux.go

```diff
@@ -0,0 +1,602 @@
+// +build linux
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package users
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/klog/v2"
+)
+
+// EntryMap holds a map of user or group entries.
+type EntryMap struct {
+	entries map[string]*entry
+}
+
+// entry is a structure that holds information about a UNIX user or group.
+// It partialially conforms parsing of both users from /etc/passwd and groups from /etc/group.
+type entry struct {
+	name      string
+	id        int64
+	gid       int64
+	userNames []string
+	shell     string
+}
+
+// limits is used to hold information about the minimum and maximum system ranges for UID and GID.
+type limits struct {
+	minUID, maxUID, minGID, maxGID int64
+}
+
+const (
+	// these are constants used when parsing /etc/passwd or /etc/group in terms of how many
+	// fields and entry has.
+	totalFieldsGroup = 4
+	totalFieldsUser  = 7
+
+	// klogLevel holds the klog level to use for output.
+	klogLevel = 5
+
+	// noshell holds a path to a binary to disable shell login.
+	noshell = "/bin/false"
+)
+
+var (
+	// these entries hold the users and groups to create as defined in:
+	// https://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane
+	usersToCreateSpec = []*entry{
+		{name: "kubeadm-etcd"},
+		{name: "kubeadm-kas"},
+		{name: "kubeadm-kcm"},
+		{name: "kubeadm-ks"},
+	}
+	groupsToCreateSpec = []*entry{
+		{name: "kubeadm-etcd", userNames: []string{"kubeadm-etcd"}},
+		{name: "kubeadm-kas", userNames: []string{"kubeadm-kas"}},
+		{name: "kubeadm-kcm", userNames: []string{"kubeadm-kcm"}},
+		{name: "kubeadm-ks", userNames: []string{"kubeadm-ks"}},
+		{name: "kubeadm-sa-key-readers", userNames: []string{"kubeadm-kas", "kubeadm-kcm"}},
+	}
+
+	// defaultLimits holds the default limits in case values are missing in /etc/login.defs
+	defaultLimits = &limits{minUID: 100, maxUID: 999, minGID: 100, maxGID: 999}
+)
+
+// ID returns the ID for an entry based on the entry name.
+// In case of a user entry it returns the user UID.
+// In case of a group entry it returns the group GID.
+// It returns false as the first returned argument if no such entry exists.
+func (u *EntryMap) ID(name string) (bool, *int64) {
```

Comment:
- once bool is removed it allows returning `nil` in case a user is missing instead of e.g. -1.
- the value can be used directly when passing to SecurityContext.RunAs* which requires *int64, instead of creating a local variable.